horovod/tensorflow/mpi_ops.cc

```diff
@@ -248,11 +276,47 @@ TFOpContext::AllocateOutput(common::TensorShape shape,
   return ConvertStatus(status);
 }
 
+int GetDeviceID(OpKernelContext* context);
+
 common::Status
 TFOpContext::AllocateZeros(int64_t num_elements, common::DataType dtype,
                            std::shared_ptr<common::Tensor>* tensor) {
-  return common::Status::PreconditionError(
-      "AllocateZeros is not supported for TensorFlow yet.");
+  ::tensorflow::Tensor* unused;
+  std::shared_ptr<PersistentTensor> zero_tensor = std::make_shared<PersistentTensor>();
+  auto tf_data_type = GetTFDataType(dtype);
+  ::tensorflow::AllocatorAttributes tf_attribute;
+  int device_ = GetDeviceID(context_); 
+  auto hvd_context = std::make_shared<TFOpContext>(context_);
+  if (device_ != CPU_DEVICE_ID){
+    tf_attribute.set_on_host(false);
+  } else{
+    tf_attribute.set_on_host(true);
+  }
+
+  Status status = context_->allocate_persistent(tf_data_type, ::tensorflow::TensorShape({num_elements}), zero_tensor.get(), &unused, tf_attribute);
+   
+  if (device_ != CPU_DEVICE_ID){
+#if HAVE_GPU
+    cudaMemset((void*)zero_tensor->AccessTensor(hvd_context->GetKernelContext())->tensor_data().data(), 0,
```

Comment:
Actually, just found something we should fix here. This should be a `cudaMemsetAsync` and pass in the stream from `device_context->stream()` or else the `cudaMemset` will launched in the default CUDA stream and cause undesired device-wide synchronization behavior. I think using this particular stream should ensure the memset occurs after the TF tensor is allocated. The call needs to be something like:
```
auto device_context = context_->op_device_context();
auto stream = (device_context != nullptr) ? AsGpuStreamValue(device_context->stream()) : 0;
cudaMemsetAsync((void*)zero_tensor->AccessTensor(hvd_context->GetKernelContext())->tensor_data().data(), 0, 
                zero_tensor->AccessTensor(hvd_context->GetKernelContext())->tensor_data().size(), stream);
```
@tgaddair can you also comment on this? I am not as well versed in this part of the code, but this seems to be the behavior we want. I'm also not quite clear on when `context_->op_device_context()` could be `nullptr` if running on GPU.

Reply:
Nice catch!  I don't think it should ever be `nullptr`, as we get the device from the [GetDeviceID](https://github.com/horovod/horovod/blob/master/horovod/tensorflow/mpi_ops.cc#L264), which always returns a valid GPU ID or the CPU ID (-1).  If we're on CPU, then we don't hit this code path, so we shouldn't need to check for that.

Maybe we can raise an error if the `device_context` is not valid?

Reply:
We could raise an error, but we also have a similar check for `nullptr` here in the existing code and would need to do the same thing: https://github.com/horovod/horovod/pull/1830/files#diff-ab99f007f05dd3de4771697e8192483aR315

Since `device_context` shouldn't ever be `nullptr` in any user case running on GPU (I assume this would only show up during code development), I don't think it needs a runtime error.

Reply:
Okay, do I have to be consistent with similar checks in existing code? Check whether device_context is nullptr

Reply:
Good question.  We should check for nullptr to be consistent.

Reply:
@romerojosh 
The AsGpuStreamValue function was introduced in tensorflow version 1.12.1: https://github.com/tensorflow/tensorflow/blob/v1.12.1/tensorflow/stream_executor/gpu/gpu_stream.h
so for < tensorflow version 1.12.1, it seems that is AsCUDAStreamValue function:
https://github.com/tensorflow/tensorflow/blob/v1.12.0/tensorflow/stream_executor/cuda/cuda_stream.h
What are your suggestions for this situation?

Reply:
Good catch.  Let's do it the same way we do it for PyTorch.  We can define a macro called `TENSORFLOW_VERSION` which we use like [this](https://github.com/horovod/horovod/blob/master/horovod/torch/ready_event.cc#L17).  This macro can be defined in setup.py like we do [here](https://github.com/horovod/horovod/blob/master/setup.py#L1272).  There's a function called [parse_version](https://github.com/horovod/horovod/blob/master/setup.py#L1007) you can use to generate the numerical representation of the version string.

Reply:
Ok, the corresponding result is: for different tensorflow versions, call their corresponding functions

Reply:
Yes, and we can do this conditional logic at compile time via macros.